File: testdata/gtscripts/genome_stream.lua
--[[
  Copyright (c) 2007 Gordon Gremme <gordon@gremme.org>
  Copyright (c) 2007 Center for Bioinformatics, University of Hamburg

  Permission to use, copy, modify, and distribute this software for any
  purpose with or without fee is hereby granted, provided that the above
  copyright notice and this permission notice appear in all copies.

  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
]]

-- testing the Lua bindings for the GenomeStream interface

function usage()
  io.stderr:write(string.format("Usage: %s testdata_dir\n", arg[0]))
  io.stderr:write("Test the GenomeStream bindings.\n")
  os.exit(1)
end

if arg[1] then
  testdata = arg[1]
else
  usage()
end

-- testing gt.gff3_in_stream_new_sorted
rval, err = pcall(gt.gff3_in_stream_new_sorted, "undefined")
assert(not rval)
assert(string.find(err, "does not exist"))

-- test correct file
gs = gt.gff3_in_stream_new_sorted(testdata.."/gff3_file_1_short.txt")
gn = gs:next_tree()
while (gn) do
  -- do something with the node...
  gn = gs:next_tree()
end

-- test corrupt file
gs = gt.gff3_in_stream_new_sorted(testdata.."/gt_gff3_fail_1.gff3")
rval, err = pcall(gs.next_tree, gs)
assert(not rval)
assert(string.find(err, "already been defined"))

-- test unsorted file
gs = gt.gff3_in_stream_new_sorted(testdata.."/unsorted_gff3_file.txt")
rval, err = pcall(gs.next_tree, gs)
assert(not rval)
assert(string.find(err, "is not sorted"))


-- test custom streams, missing override
cs = gt.custom_stream_new_sorted()
rval, err = pcall(cs.next_tree, cs)
assert(not rval)
assert(string.find(err, "method defined in custom stream"))

-- test custom streams, wrong return type
gs = gt.gff3_in_stream_new_sorted(testdata.."/eden.gff3")
cs = gt.custom_stream_new_sorted()
function cs:next_tree()
  if gs:next_tree() then
    return 1
  end
end
rval, err = pcall(gs.next_tree, cs)
assert(not rval)
assert(string.find(err, "return a genome node"))

-- test custom streams, runtime error
gs = gt.gff3_in_stream_new_sorted(testdata.."/eden.gff3")
cs = gt.custom_stream_new_sorted()
function cs:next_tree()
  n = gs:next_tree()
  x = 1 + nil
  return n
end
rval, err = pcall(gs.next_tree, cs)
assert(not rval)
assert(string.find(err, "perform arithmetic on a nil"))

-- test custom streams
gs = gt.gff3_in_stream_new_sorted(testdata.."/eden.gff3")
cs = gt.custom_stream_new_sorted()
cs.instream = gs
function cs:next_tree()
  return cs.instream:next_tree()
end
rval, err = pcall(gs.next_tree, cs)
assert(rval)

-- test custom streams
cs = gt.custom_stream_new_sorted()
cs.count = 1
function cs:next_tree()
  if cs.count <= 5 then
    n = gt.feature_node_new("test", "gene", 1+cs.count, 100+cs.count, "+")
    cs.count = cs.count + 1
  else
    n = nil
  end
  return n
end
nodes = {}
local gn = cs:next_tree()
while (gn) do
  table.insert(nodes, gn)
  gn = cs:next_tree()
end
assert(#nodes == 5)

File: src/extended/genome_stream.c
/*
  Copyright (c) 2006-2008 Gordon Gremme <gremme@zbh.uni-hamburg.de>
  Copyright (c) 2006-2008 Center for Bioinformatics, University of Hamburg

  Permission to use, copy, modify, and distribute this software for any
  purpose with or without fee is hereby granted, provided that the above
  copyright notice and this permission notice appear in all copies.

  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/

#include <assert.h>
#include <stdarg.h>
#include "core/ma.h"
#include "core/unused_api.h"
#include "extended/genome_stream_rep.h"

GenomeStream* genome_stream_create(const GenomeStreamClass *gsc,
                                   bool ensure_sorting)
{
  GenomeStream *gs;
  assert(gsc && gsc->size);
  gs = gt_calloc(1, gsc->size);
  gs->c_class = gsc;
  gs->ensure_sorting = ensure_sorting;
  return gs;
}

GenomeStream* genome_stream_ref(GenomeStream *gs)
{
  assert(gs);
  gs->reference_count++;
  return gs;
}

void genome_stream_delete(GenomeStream *gs)
{
  if (!gs) return;
  if (gs->reference_count) {
    gs->reference_count--;
    return;
  }
  assert(gs->c_class);
  if (gs->c_class->free) gs->c_class->free(gs);
  gt_genome_node_rec_delete(gs->buffer);
  gt_free(gs);
}

int genome_stream_next_tree(GenomeStream *gs, GT_GenomeNode **gn, GT_Error *err)
{
  GT_GenomeNode *new_node = NULL;
  int had_err = 0;
  assert(gs && gs->c_class && gs->c_class->next_tree);
  gt_error_check(err);
  /* filling */
  if (!gs->buffer)
    had_err = gs->c_class->next_tree(gs, &gs->buffer, err);
  if (!had_err && gs->buffer)
    had_err = gs->c_class->next_tree(gs, &new_node, err);
#ifndef NDEBUG
  /* checking */
  if (!had_err && gs->ensure_sorting && gs->buffer && new_node) {
    assert(gt_genome_node_compare(&gs->buffer, &new_node) <= 0);
  }
#endif
  /* serving */
  if (!had_err) {
    *gn = gs->buffer;
    gs->buffer = new_node;
  }
  return had_err;
}

bool genome_stream_is_sorted(GenomeStream *gs)
{
  assert(gs);
  return gs->ensure_sorting;
}

void* genome_stream_cast(GT_UNUSED const GenomeStreamClass *gsc, GenomeStream *gs)
{
  assert(gsc && gs && gs->c_class == gsc);
  return gs;
}
